utils/mil_tools/mil_ros_tools/labeled_bag_tester.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python
+
+import rosbag
+import rospy
+import roslib
+import time
+import cv2
+import numpy as np
+from sensor_msgs.msg import CameraInfo
+
+class LabeledBagTester(object):
+    '''
+    A utility class to play a bag containing mil_msgs/LabeledObjects
+    messages, calling a particular function every time one of these
+    messages is encountered. This is intended to be used by perception
+    integration tests.
+    '''
+    def __init__(self):
+        self._pubs = {}
+        self._cur_bag = None
+
+    def _setup_publishers(self):
+        '''
+        Internal function. Creates a publisher for each topic in a bag
+        to be used during playback
+        '''
+        self._pubs = {}
+        types, topics = self._cur_bag.get_type_and_topic_info()
+        for t, info in topics.iteritems():
+            msg_class = roslib.message.get_message_class(info.msg_type)
+            if msg_class is not None:
+                self._pubs[t] = rospy.Publisher(t, msg_class, queue_size=3)
+
+    @staticmethod
+    def polygon_to_contour(polygon):
+        '''
+        Static method to convert the polygon in a label to a (n x 2)
+        numpy array.
+        '''
+        arr = np.zeros((len(polygon), 2))
+        for i, p in enumerate(polygon):
+            arr[i][0] = p.x
+            arr[i][1] = p.y
+        return arr
+
+    def test_bag(self, filename):
+        '''
+        Calls self.start() then plays back the bag file specifed in
+        filename, calling self.verify(msg) every time a LabeledObjects
+        message appears. Finnaly calls self.stop() once the bag is done.
+        '''
+        self._cur_bag = rosbag.Bag(filename)
+        self._setup_publishers()
+        _, _, last_time = self._cur_bag.read_messages(topics=self._pubs.keys()).next()
+        deficit = 0.0 # Track how far behind we are in playback
+        self.start()
+        for topic, msg, t in self._cur_bag.read_messages(topics=self._pubs.keys()):
+            if msg._type == 'mil_msgs/LabeledObjects':
+                self.verify(msg)
+            x = time.time()
+            if rospy.is_shutdown():
+                return
+            dt = t - last_time
+            last_time = t
+            self._pubs[topic].publish(msg)
+            dx = time.time() - x
```

Comment:
Why not use rostime? This prevents you from running this at different bag playback speeds.